packages/indexer-agent/src/network-subgraph.ts

```diff
@@ -0,0 +1,251 @@
+  }
+
+  private constructor(options: NetworkSubgraphOptions) {
+    this.logger = options.logger
+
+    if (options.endpoint) {
+      this.endpointClient = createClient({
+        url: options.endpoint,
+        fetch,
+        requestPolicy: 'network-only',
+      })
+    }
+
+    if (options.deployment) {
+      const client = createClient({
+        url: new URL(
+          `/subgraphs/id/${options.deployment.id.ipfsHash}`,
+          options.graphNodeQueryEndpoint,
+        ).toString(),
+        fetch,
+        requestPolicy: 'network-only',
+      })
+      const status = options.deployment.status
+
+      this.deployment = {
+        id: options.deployment.id,
+        client,
+        status,
+      }
+    }
+  }
+
+  public static async create({
+    logger: parentLogger,
+    indexer,
+    graphNodeQueryEndpoint,
+    endpoint,
+    deployment,
+  }: NetworkSubgraphCreateOptions): Promise<NetworkSubgraph> {
+    // Either an endpoint or a deployment needs to be provided; the CLI
+    // validation should already guarantee that but we're asserting this again
+    // here, just to be on the safe side
+    console.assert(endpoint || deployment)
+
+    const logger = parentLogger.child({
+      component: 'NetworkSubgraph',
+      endpoint,
+      deployment: deployment?.ipfsHash,
+    })
+
+    let deploymentInfo:
+      | { id: SubgraphDeploymentID; status: Eventual<DeploymentStatus> }
+      | undefined
+
+    if (deployment) {
+      // Make sure the network subgraph is being indexed
+      await indexer.ensure(
+        `${deployment.ipfsHash.slice(0, 23)}/${deployment.ipfsHash.slice(23)}`,
+        deployment,
+      )
+
+      const status = await monitorDeployment({
+        logger,
+        indexer,
+        deployment,
+      })
+
+      deploymentInfo = {
+        id: deployment,
+        status,
+      }
+    }
+
+    // Create the network subgraph instance
+    const networkSubgraph = new NetworkSubgraph({
+      logger,
+      graphNodeQueryEndpoint,
+      endpoint,
+      deployment: deploymentInfo,
+    })
+
+    // If we don't have a network subgraph endpoint configured, we
+    // need to wait until the deployment is synced
+    if (!endpoint) {
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      await deploymentInfo!.status.filter(status => status.synced).value()
+    }
+
+    return networkSubgraph
+  }
+
+  private async getClient(): Promise<Client> {
+    if (this.deployment) {
+      const status = await this.deployment.status.value()
+      const healthy = status.synced && status.health === 'healthy'
+
+      if (healthy) {
+        return this.deployment.client
+      } else if (this.endpointClient) {
+        return this.endpointClient
+      } else {
+        // We have no endpoint and our deployment is not synced or unhealthy;
+        // we could crash here or we are more optimistic and query the deploymet
+        // for now
+        return this.deployment.client
```

Comment:
The one reason I don't want to do that is that we might be logging this every 10s. Maybe crashing _is_ the best solution, because a stuck network subgraph doesn't help either.